Spatial_Engine: Add support for tapered flange sections

JosefTaylor opened the pull request:
<!-- PLEASE ENSURE YOU REVIEW THE CONTENT OF EACH PR CAREFULLY, INCLUDING SUBSEQUENT COMMENTS BY YOURSELF OR OTHERS. -->
<!-- IN PARTICULAR PLEASE ENSURE THAT SENSITIVE OR INAPPROPRIATE INFORMATION IS NOT UPLOADED -->

### NOTE: Depends on 
<!-- Link to any additional PRs in other repos required for this PR to function -->
<!-- Delete if not required -->
[BHoM 1316](https://github.com/BHoM/BHoM/pull/1316)
   
### Issues addressed by this PR
<!-- Add reference(s) to issue(s) solved by this PR. Please use keyword Fixes/Closes as per https://help.github.com/articles/closing-issues-using-keywords/ -->

Closes #

<!-- Add short description of what has been fixed -->


### Test files
<!-- Link to test files to validate the proposed changes -->
https://burohappold.sharepoint.com/:f:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/BHoM/Spatial_oM/%231316-add-tapered-flange-sections?csf=1&web=1&e=aMhwGq

### Changelog
<!-- Text to go into changelog if applicable -->
<!-- Please see https://github.com/BHoM/documentation/wiki/changelog for guidelines -->


### Additional comments
<!-- As required -->

IsakNaslundBh commented:
Think this generally looks great!

a small thing is to centre the channels around the origin. THere is some code to help with that for the parallel flange channels.

Also, would be good to get Torsion and Warping constant calculations included as well. Have found some formulae for the Torsion constants for both sections types

JosefTaylor commented:
@IsakNaslundBh From what I saw, it looks like the profiles auto-position themselves at their centroids, or at least they do when you preview them in grasshopper! The parallel flange I section is drawn with the bottom center at the origin, and is not translated based on its centroid.

Should I get torsional calculations in this one? That applies to all tees and channels, so I figured it would be a separate PR

I'll change the tag- wasn't sure if adapting to upstream changes (in BHoM) would be feature or compliance

JosefTaylor commented:
Also- I sort of hid the Fillet() methods as private methods; where is the right place for those?

JosefTaylor commented:
Hey @IsakNaslundBh I added the torsional constants, I'd love a second set of eyes on this as I've added a few methods; perhaps we can clean things up further.

IsakNaslundBh reviewed (changes requested):
Have made a test of the torsional constant now, and unfortunately, it fails for Standard UPN sections, as they are outside the interpolate range of the interpolation for the EndLossCorrection. For the IPN beams the values of the torsional constant is quite conservative compared to tabulated values Good to chat this through I think!

I have added a test script with all IPN and UPN beams created here: 
https://burohappold.sharepoint.com/:f:/s/BHoM/EhqVN6PD5r5PvPQhbWslSqIB6F-DelaGw1WzG1dicm2qTw?e=MgKV74

and [here](https://www.dlubal.com/en/cross-section-properties/series-ipn-en-10365-2017-arcelormittal-2018,series-upn-en-10365-2017-arcelormittal-2018) is where I got the data from that could be used as an initial check.

Also, some details:
- InputFromProperty for flangeSlope is missing on the tapered I section.
- Profiles still need to be centroid aligned (for the purpose of using them on the physical elements)

In general I like the route of the new helper methods added, think that is a good idea and makes sense in general! 👍 

Probably worth a quick call next week to run through a bit!

IsakNaslundBh commented on Structure_Engine/Query/TorsionalConstant.cs:
```diff
@@ -406,6 +497,118 @@ private static double AlphaLJunction(double tw, double tf, double r)
+        {
+            List<double> slopes = new List<double>
+            {
+                Math.Atan(0),
+                Math.Atan(1 / 50),
+                Math.Atan(1 / 20),
+                Math.Atan(1 / 6),
```
Also, forgot to add in the review, these all evaluate to 0 as it is assuming integer division 🤦 
Adding a `.0` at the end should help fix it.
```suggestion
                Math.Atan(0),
                Math.Atan(1 / 50.0),
                Math.Atan(1 / 20.0),
                Math.Atan(1 / 6.0),
```

IsakNaslundBh commented on Structure_Engine/Query/TorsionalConstant.cs:
```diff
@@ -406,6 +497,118 @@ private static double AlphaLJunction(double tw, double tf, double r)
+            //Equation 28
+            double alpha16 = -0.0836 + 0.2536 * tw / t2 + 0.1268 * r / t2 - 0.0806 * tw * r / Math.Pow(t2, 2) - 0.0858 * Math.Pow(tw / t2, 2); 
+            //Interpolate alpha between parallel flange and slope of 1 to 6:
+            return alpha0 + alpha16 * s / Math.Atan(0.1667);
```
Think this interpolation is a bit off. Think it should be the following:
```suggestion
            return alpha0 + (alpha16 - alpha0) * s / Math.Atan(0.1667);
```

IsakNaslundBh commented on Structure_Engine/Query/TorsionalConstant.cs:
```diff
@@ -406,6 +497,118 @@ private static double AlphaLJunction(double tw, double tf, double r)
+            //Equation 30
+            double alpha16 = -0.1325 + 0.3015 * tw / t2 + 0.1400 * r / t2 - 0.1070 * tw * r / Math.Pow(t2, 2) - 0.0956 * Math.Pow(tw / t2, 2); 
+            //Interpolate alpha between parallel flange and slope of 1 to 6:
+            return alpha0 + alpha16 * s / Math.Atan(0.1667);
```
as above:
```suggestion
            return alpha0 + (alpha16 - alpha0) * s / Math.Atan(0.1667);
```

IsakNaslundBh commented on Structure_Engine/Query/TorsionalConstant.cs:
```diff
@@ -385,7 +448,35 @@ private static double InscribedDiameterLJunction(double tw, double tf, double r)
+        private static double InscribedDiameterTaperTJunction(double tw, double t3, double r, double s)
+        {
+            //Equation 24
+            double f = r * s * (Math.Sqrt((1 / Math.Pow(s, 2)) + 1) - 1 - tw / (2 * r));
```
good to check for s == 0 here. For that case we could just return the non taper inscribed diameter

IsakNaslundBh reviewed (commented):
Some more small comments on things spotted, and as discussed to change for using the tables to using the formulation for the end correction bit :)

JosefTaylor commented:
Alright, I made a much more useful test script (check parametric parallel vs tapered sections) and fixed a few more errors, which finally brings it to agreement between a tapered section with zero slope and a parallel section.

Also, checked US taper flange I and channel sections, and the calculated J agrees with the published J on average 4% up to 13% at worst. (this one is an outlier, everything else is much closer)

JosefTaylor commented:
@BHoMBot check required

IsakNaslundBh commented:
@BHoMBot check code-compliance

IsakNaslundBh commented on Structure_Engine/Query/TorsionalConstant.cs:
```diff
@@ -355,57 +418,152 @@ public static double ITorsionalConstant(this IProfile profile)
+            if (0.2 > tw / t2 || tw / t2 > 1.0)
+                Reflection.Compute.RecordWarning("Calculation of alpha term of torsional constant may not be accurate because ratio of web thickness to flange thickness is out of the applicable range.");
+
+            double alpha0 = AlphaTJunction(tw, t2, r);
```
Think `tf` should be used here

```suggestion
            double alpha0 = AlphaTJunction(tw, tf, r);
```

IsakNaslundBh commented on Structure_Engine/Query/TorsionalConstant.cs:
```diff
@@ -355,57 +418,152 @@ public static double ITorsionalConstant(this IProfile profile)
+        [Input("s", "Flange taper slope.", typeof(Ratio))]
+        public static double InscribedDiameterTaperTJunction(double tw, double t3, double r, double s)
+        {
+            if (s < 0.001) return InscribedDiameterTJunction(tw, t3, r);
```
Could make this tolerance significantly smaller, even if that will not matter that much ofc. Maybe use `Tolerance.Angle` instead (I know it is not an angle, but is the closest). Good to avoid using hardcoded tolerances like this unless we have to, Good to also put a short comment why it is done if we come back to this later as well.

JosefTaylor replied:
```suggestion
            if (s < Tolerance.Angle) return InscribedDiameterTJunction(tw, t3, r);
```

IsakNaslundBh reviewed (changes requested):
One small tweak more for one of the alphas and some small comments, then I think this is looking good!

Getting values that vary slightly from tabulated values, but formulas used are looking correct to this well cited source, so I am happy with it, Great job in general.

Also, have found a citation that the same formulas that are used for non-tapered sections can be used for tapered for warping constants, but that can be added in a subsequent PR.

JosefTaylor commented:
@BHoMBot check required

IsakNaslundBh commented on Spatial_Engine/Create/ShapeProfiles/TaperFlangeISectionProfile.cs:
```diff
@@ -0,0 +1,188 @@
+                return null;
+            }
+
+            if (toeRadius > flangeThickness - width / 4 * Math.Tan(flangeSlope))
```
Think it should be like this now?
```suggestion
            if (toeRadius > flangeThickness - width / 4 * flangeSlope)
```

IsakNaslundBh commented on Spatial_Engine/Create/ShapeProfiles/TaperFlangeISectionProfile.cs:
```diff
@@ -0,0 +1,188 @@
+
+            if (toeRadius > flangeThickness - width / 4 * Math.Tan(flangeSlope))
+            {
+                InvalidRatioError("toeRadius", "flangeThickness");
```
```suggestion
                InvalidRatioError("toeRadius", "flangeThickness and flangeSlope");
```

IsakNaslundBh commented on Spatial_Engine/Create/ShapeProfiles/TaperFlangeISectionProfile.cs:
```diff
@@ -0,0 +1,188 @@
+        [Output("I", "The created ISectionProfile.")]
+        public static TaperFlangeISectionProfile TaperFlangeISectionProfile(double height, double width, double webThickness, double flangeThickness, double flangeSlope, double rootRadius, double toeRadius)
+        {
+            if (height < flangeThickness * 2 + rootRadius * 2 || height <= flangeThickness * 2)
```
This need to take the flange slope into account, basically checking the thickness next to the web, and then find a way to add the radius. Not sure it should only be the root radius though, think that is also depending on the slope.

IsakNaslundBh commented on Spatial_Engine/Create/ShapeProfiles/TaperFlangeISectionProfile.cs:
```diff
@@ -0,0 +1,188 @@
+        {
+            if (height < flangeThickness * 2 + rootRadius * 2 || height <= flangeThickness * 2)
+            {
+                InvalidRatioError("height", "flangeThickness and rootRadius");
```
this should mention the slope for the error message

IsakNaslundBh commented on Spatial_Engine/Create/ShapeProfiles/TaperFlangeISectionProfile.cs:
```diff
@@ -0,0 +1,188 @@
+                return null;
+            }            
+
+            if (flangeSlope > Math.Atan(4 * flangeThickness / width))
```
Do not think this should be Atan any longer?

IsakNaslundBh commented on Spatial_Engine/Create/ShapeProfiles/TaperFlangeChannelProfile.cs:
```diff
@@ -0,0 +1,140 @@
+        [Output("channel", "The created ChannelProfile.")]
+        public static TaperFlangeChannelProfile TaperFlangeChannelProfile(double height, double flangeWidth, double webThickness, double flangeThickness, double flangeSlope, double rootRadius = 0, double toeRadius = 0, bool mirrorAboutLocalZ = false)
+        {
+            if (height < flangeThickness * 2 + rootRadius * 2 || height <= flangeThickness * 2)
```
Also need to account for the slope

IsakNaslundBh commented on Spatial_Engine/Create/ShapeProfiles/TaperFlangeChannelProfile.cs:
```diff
@@ -0,0 +1,140 @@
+                return null;
+            }
+
+            if (flangeSlope > Math.Atan(2 * flangeThickness / flangeWidth))
```
Atan?

IsakNaslundBh commented on Spatial_Engine/Create/ShapeProfiles/TaperFlangeChannelProfile.cs:
```diff
@@ -0,0 +1,140 @@
+                return null;
+            }
+
+            if (toeRadius > flangeThickness - flangeWidth/2 * Math.Tan(flangeSlope) )
```
Tan?

IsakNaslundBh commented on Spatial_Engine/Create/ShapeProfiles/TaperFlangeChannelProfile.cs:
```diff
@@ -0,0 +1,140 @@
+
+            if (toeRadius > flangeThickness - flangeWidth/2 * Math.Tan(flangeSlope) )
+            {
+                InvalidRatioError("flangeThickness", "toeRadius");
```
All parameters to be mentioned

IsakNaslundBh commented on Spatial_Engine/Create/ShapeProfiles/TaperFlangeChannelProfile.cs:
```diff
@@ -0,0 +1,140 @@
+        {
+            if (height < flangeThickness * 2 + rootRadius * 2 || height <= flangeThickness * 2)
+            {
+                InvalidRatioError("height", "flangeThickness and rootRadius");
```
Flange slope to be mentioned

IsakNaslundBh reviewed (changes requested):
SOme more create method checks required

JosefTaylor commented:
@BHoMBot  check required

IsakNaslundBh commented:
@BHoMBot check copyright-compliance
@BHoMBot check dataset-compliance

IsakNaslundBh reviewed (approved):
LGTM now!

IsakNaslundBh commented:
@BHoMBot check ready-to-merge
